### ISSUE TITLE
G115 is reporting false positives (a summary)

### ISSUE THREAD

**Issue by czechbol** [securego/gosec#1212]
### Summary 
G115 is still reporting false positives even after #1189 and #1194.

### Known false positives
any kind of arithmetic done on the checked variable after the check:
```go
func foo(a int) uint {
    if a != 3 || a != 4 {
        panic("not supported")
    }
    return uint(a-1)
}
```
binary truncation (reported by @stephenc):
```go
func foo(a int) uint16 {
    return uint16(a &0xffff)
}
```
builtin min and max functions (reported by @ben-krieger):
```go
func foo() uint16 {
    a, b := 1234, 2345
    res := min(a,b)

    return uint(result)
}
```

### Expected behavior
These shouldn't raise any warnings.

### Actual behavior
They raise G115 warnings.

### Notes:
Do not attempt to implement this as a single PR, the changes necessary to support everything will be pretty complex. It will be much better to open small PRs that fix one issue at a time.

**Comment by ccoVeille** [comment]
At least these use cases are less frequent than the previous ones.

**Comment by deerbone** [comment]
I'd say that people will hit the first one pretty frequently, and it also seems to me it's the most difficult to tackle without bringing in any false negatives.

**Comment by Victoremepunto** [comment]
I think I may be hitting this: https://github.com/RedHatInsights/frontend-operator/pull/188/files#diff-e8b328ffb5ced885894c80b942846601597e05020f831c2b1442d8fb8a654c8dR120

**Comment by czechbol** [comment]
@Victoremepunto you're getting that warning since golangci-lint has not released a new version with the changes from #1194.

@ccoVeille do you know and can share at least approximately when we can expect it?

**Comment by ccoVeille** [comment]
I'm just a random gopher. I'm pretty active, but I'm not part of golangci-lint release.

@ldez might be a better candidate to reply that question

**Comment by PlasmaPower** [comment]
I'm not sure if I should open a separate issue for this, but another false positive is casting the index produced by iterating over a range to an unsigned int. We do this a lot in our codebase. E.g.:

```go
myArr := []string{"hello", "world"}

for i := range len(myArr) {
    _ = uint64(i) // shouldn't error but does with G115
}

for i, _ := range myArr {
    _ = uint64(i) // shouldn't error but does with G115
}
```

Ideally it'd be able to look at bounds in for loops as well, but I get that's more complicated:

```go
for i := 0; i < 10; i++ {
    _ = uint64(i) // shouldn't error but does with G115
}

for i := randIntMightBeNegativeEven(); i >= 0; i-- {
    _ = uint64(i) // shouldn't error but does with G115
}
```

**Comment by ccoVeille** [comment]
Quick follow up, [golangci-lint 1.61.0](https://github.com/golangci/golangci-lint/releases/tag/v1.61.0) was shipped with [gosec 2.21.2](https://github.com/golangci/golangci-lint/pull/4997) :muscle: 

So everything that was made in gosec to improve G115 detection are now released :tada:

**Comment by czechbol** [comment]
> I'm not sure if I should open a separate issue for this, but another false positive is casting the index produced by iterating over a range to an unsigned int. We do this a lot in our codebase. E.g.:

@PlasmaPower You are correct, this seems like another issue. I'll get back to it a bit later and update my summary.

**Comment by Victoremepunto** [comment]
> @Victoremepunto you're getting that warning because golangci-lint has not released a new version with the changes from #1194 yet. The code you have there looks sound so once we have a new golangci-lint release, it should work.
> 
> @ccoVeille do you know and can share at least approximately when we can expect it?

Hi! unfortunately, using golangci-lint 1.60.1 the problem is still there - might worth trying to get a simpler reproducer and add it to this issue? 

check this: https://github.com/RedHatInsights/frontend-operator/actions/runs/10835444044/job/30066989278?pr=188#step:5:26

**Comment by ldez** [comment]
> Hi! unfortunately, using golangci-lint 1.60.1 the problem is still there

Maybe it's because the fix is inside golangci-lint v1.61.0 and not v1.60.1 :wink:

**Comment by Victoremepunto** [comment]
> > Hi! unfortunately, using golangci-lint 1.60.1 the problem is still there
> 
> 
> 
> Maybe it's because the fix is inside golangci-lint v1.61.0 and not v1.60.1 :wink: 

lol, I'm sorry the version I used was v1.61.0, I just mistyped it. Here's an excerpt of the GHA logs:


```

run golangci-lint
  Running [/home/runner/golangci-lint-1.61.0-linux-amd64/golangci-lint run] in [/home/runner/work/frontend-operator/frontend-operator] ...
  Error: controllers/status.go:120:44: G115: integer overflow conversion int -> int32 (gosec)

```

**Comment by czechbol** [comment]
Thanks @ben-krieger for the explanation. I purposefully decided to ignore TOCTOU in that PR. 

The proposal of a separate lint rule for it is a very good idea since a different lint message would explain the issue better.

**Comment by ldemailly** [comment]
Contributing a real world example of false positive with min:
```go
			p1.R = uint8(min(255, uint16(p1.R)+uint16(p2.R)))
```

**Comment by ccojocar** [comment]
On this, I think we need to start the __Obfuscated Go Integer Overflow Contest__. 😄

**Comment by MichalSalon** [comment]
Another one I stumbled upon after updating golangci-lint to 1.61.0
```go
func foo(hash string) (uint32, error) {
	decodedHash, err := base64.RawStdEncoding.DecodeString(hash)
	if err != nil {
		return 0, err
	}

	if hashLen := len(decodedHash); hashLen > math.MaxUint32 {
		return 0, fmt.Errorf("invalid decoded hash length %d, exceeds max value for uint32", hashLen)
	}
	return uint32(len(decodedHash)), nil
}
```

It works if I change it to:
```go
func foo(hash string) (uint32, error) {
	decodedHash, err := base64.RawStdEncoding.DecodeString(hash)
	if err != nil {
		return 0, err
	}

	hashLen := len(decodedHash)
	if hashLen > math.MaxUint32 {
		return 0, fmt.Errorf("invalid decoded hash length %d, exceeds max value for uint32", hashLen)
	}
	return uint32(hashLen), nil
}
```

But since the slice was created inside of the function and is not changed in any way, I would still say that this is a false positive.

**Comment by ctarsjp** [comment]
Another false positive (similar to binary truncation with '&', but using bit shifting):

```go
var (
	u32 uint32
	u8  uint8
)

u8 = uint8(u32 >> 24)
```

**Comment by hairyhenderson** [comment]
I ran into a false G115 positive just now, using `strconv.ParseUint`:

```go
func main() {
	a := "13"
	b, _ := strconv.ParseUint(a, 10, 8)
	c := int(b)
	fmt.Printf("%d\n", c)
}
```

This fails with `G115: integer overflow conversion uint64 -> int`

There's actually [a unit test](https://github.com/securego/gosec/blob/master/testutils/g115_samples.go#L414-L415) similar to this issue, except it casts to a `uint8` rather than an `int`.

**Comment by hairyhenderson** [comment]
I've issued https://github.com/securego/gosec/pull/1254 to fix the `ParseUint` issue I described above.

**Comment by theory** [comment]
Another false positive with `min`:

``` go
func main() {
	if len(os.Args) > 1 {
		num, err := strconv.ParseInt(os.Args[1], 10, 64)
		if err != nil { panic(err) }
		var x uint32
		if num >= 0 {
			x = uint32(min(num, math.MaxUint32))
		}
		fmt.Printf("%v\n", x)
	}
}
```

Can only avoid the issue by changing the `if` statement to

``` go
		if num >= 0 && num <= math.MaxUint32 {
			x = uint32(num)
		}
```

**Comment by ad-zsolt-imre** [comment]
Golang version: `1.23.5`
Gosec version: `v2.22.1`


The below also reports "integer overflow conversion".

```go
uint64(time.Now().UnixMilli())
```

Given max `uint64` (`18446744073709551615`) is greater than max `int64` (`9223372036854775807`), how can a conversion from `int64` to `uint64` result in an "overflow"? Wouldn't the opposite, `uint64` -> `int64` result in an overflow if the value of `uint64` greater than max `int64`?

**Comment by kaovilai** [comment]
```
pkg/repository/udmrepo/kopialib/lib_repo.go:527:17: G115: integer overflow conversion int -> int32 (gosec)
                Length:  int32(min(mani.Length, math.MaxInt32)),
```
another false postive.

**Comment by theory** [comment]
Here's another one:

```go
const multiple = 4
need := uint8(multiple - (len(digest) % multiple))
```

**Comment by eest** [comment]
I seem to be getting problems with nested structs, here is a contrived example:
```
package main

import (
	"fmt"
	"math"
)

type nestedStruct struct {
	i *innerStruct
}

type innerStruct struct {
	u32 *uint32
}

func main() {
	n := nestedStruct{
		i: &innerStruct{
			u32: new(uint32),
		},
	}

	i := innerStruct{
		u32: new(uint32),
	}

	var input uint32
	_, err := fmt.Scanf("%d", &input)
	if err != nil {
		panic(err)
	}

	n.i.u32 = &input
	i.u32 = &input

	if *n.i.u32 > math.MaxInt32 {
		panic("out of range")
	} else {
		i32 := int32(*n.i.u32)
		fmt.Println(i32)
	}

	if *i.u32 > math.MaxInt32 {
		panic("out of range")
	} else {
		i32 := int32(*i.u32)
		fmt.Println(i32)
	}
}
```

This gives me:
```
$ gosec ./...
[...]
[.../main.go:39] - G115 (CWE-190): integer overflow conversion uint32 -> int32 (Confidence: MEDIUM, Severity: HIGH)
    38: 	} else {
  > 39: 		i32 := int32(*n.i.u32)
    40: 		fmt.Println(i32)
```

So it seems happy with the the bounds check on `*i.u32`, but not with `*n.i.u32`

**Comment by ccojocar** [comment]
fixed by #1460

**Comment by cyphar** [comment]
@ccojocar Are you sure? #1460 appears to be fixing #1211 (not this issue -- #1212) but the description references #1212 seemingly by accident.

**Comment by cyphar** [comment]
I just hit the following example of what appears to be a false positive as well:

```go
func ProcPid(pid int) ProcBase {
	if pid < 0 || uint64(pid) >= 1<<31 {
		panic("invalid ProcBasePid value") // TODO: should this be an error?
	}
	return ProcBase{inner: libpathrs.ProcPid(uint32(pid))}
}
```

If you remove the `uint64` cast, gosec doesn't produce an error, but if you include it then you get a potential overflow error. This seems similar to the first example given in the issue description. (The `uint64` cast is necessary for the code to compile on 32-bit machines.)

This error happens with golangci-lint v2.7.2.

**Comment by ldemailly** [comment]
there are still a bunch of false positive (as off the version bundled with golanglintci 2.10.1) - should we open new issues?

edit: filled #1577 and will file more as I find them (though I would need first to just remove a bunch of //nolint:gosec )
